controllers/quay/quayregistry_controller.go

```diff
@@ -120,10 +120,91 @@ func (r *QuayRegistryReconciler) Reconcile(ctx context.Context, req ctrl.Request
+	if migrationInProgress {
+		log.Info("++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++")
+		log.Info("migration currently in progress!")
+		log.Info(fmt.Sprintf("currentVersion is: '%s'", updatedQuay.Status.CurrentVersion))
+
+		log.Info("checking Quay upgrade `Job` completion")
+
+		var upgradeJob batchv1.Job
+		err := r.Client.Get(
+			ctx,
+			types.NamespacedName{
+				Name:      updatedQuay.GetName() + "-" + v1.QuayUpgradeJobName,
+				Namespace: updatedQuay.GetNamespace(),
+			},
+			&upgradeJob,
+		)
+		if err != nil {
+			log.Error(err, "could not retrieve Quay upgrade `Job`")
+			return ctrl.Result{RequeueAfter: time.Second * 1}, nil
+		}
+
+		if upgradeJob.Status.Active == 1 {
+			return ctrl.Result{RequeueAfter: time.Second * 1}, nil
+		}
+
+		if upgradeJob.Status.Succeeded == 1 {
+			log.Info("Quay upgrade complete, updating `status.currentVersion`")
+
+			var freshQuay v1.QuayRegistry
+			if err := r.Client.Get(ctx, req.NamespacedName, &freshQuay); err != nil {
+				log.Error(err, "could not retrieve QuayRegistry")
+				return ctrl.Result{RequeueAfter: time.Second * 1}, nil
+			}
+			qcopy := freshQuay.DeepCopy()
+
+			msg := "All registry components created"
+			condition := v1.Condition{
+				Type:               v1.ConditionComponentsCreated,
+				Status:             metav1.ConditionTrue,
+				Reason:             v1.ConditionReasonComponentsCreationSuccess,
+				Message:            msg,
+				LastUpdateTime:     metav1.Now(),
+				LastTransitionTime: metav1.Now(),
+			}
+			qcopy.Status.Conditions = v1.SetCondition(qcopy.Status.Conditions, condition)
+			qcopy.Status.CurrentVersion = v1.QuayVersionCurrent
+			r.EventRecorder.Event(qcopy, corev1.EventTypeNormal, string(v1.ConditionReasonHealthChecksPassing), msg)
+
+			if err = r.Client.Status().Update(ctx, qcopy); err != nil {
+				log.Error(err, "could not update QuayRegistry status with current version")
+				return ctrl.Result{RequeueAfter: time.Second * 1}, nil
+			}
+
+			log.Info(fmt.Sprintf("currentVersion updated to: '%s'", qcopy.Status.CurrentVersion))
+
+			log.Info("successfully updated `status` after Quay upgrade")
+
+			return ctrl.Result{RequeueAfter: time.Second * 1}, nil
+		}
+
+		// when the job fails, the next reconciliation to run will think
+		// that migrations are currently not running, since we change
+		// the condition reason to v1.ConditionReasonMigrationsFailed.
+		// this doesn't necessarily describe reality, because kube will
+		// retry failed jobs for us, but it's desired behaviour because
+		// when the migration job fails due to misconfiguration, then the
+		// reconcile function should be allowed to proceed.
+		if upgradeJob.Status.Failed == 0 {
+			updatedQuay, err = r.updateWithCondition(
+				updatedQuay,
+				v1.ConditionComponentsCreated,
+				metav1.ConditionFalse,
+				v1.ConditionReasonMigrationsFailed,
+				"failed to run migrations")
```

Comment:
Nice to have: wouldn't it be cool if we could show up the Reason why it is failing? I think that job.Status.Conditions has a Failed condition with the Reason.

Reply:
That's a great idea! Its got both a reason and a message. I chose to go with the message since we already have a reason (`ConditionReasonMigrationsFailed`).
From `batchv1.JobCondition` type:
```
// [...]
	// (brief) reason for the condition's last transition.
	// +optional
	Reason string `json:"reason,omitempty" protobuf:"bytes,5,opt,name=reason"`
	// Human readable message indicating details about last transition.
	// +optional
	Message string `json:"message,omitempty" protobuf:"bytes,6,opt,name=message"`
}
```